lib/api/multiObjectDelete.js

```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
     let numOfObjectsRemoved = 0;
     const skipError = new Error('skip');
     const objectLockedError = new Error('object locked');
+    let deleteFromStorage = [];
+    const initialStep = initializeMultiObjectDeleteWithBatchingSupport(bucketName, inPlay, log);
 
-    // doing 5 requests at a time. note that the data wrapper
-    // will do 5 parallel requests to data backend to delete parts
-    return async.forEachLimit(inPlay, 5, (entry, moveOn) => {
-        async.waterfall([
-            callback => {
-                let decodedVersionId;
-                if (entry.versionId) {
-                    decodedVersionId = entry.versionId === 'null' ?
-                        'null' : versionIdUtils.decode(entry.versionId);
-                }
-                if (decodedVersionId instanceof Error) {
-                    monitoring.promMetrics('DELETE', bucketName, 404,
-                        'multiObjectDelete');
-                    return callback(errors.NoSuchVersion);
-                }
-                return callback(null, decodedVersionId);
-            },
-            // for obj deletes, no need to check acl's at object level
-            // (authority is at the bucket level for obj deletes)
-            (versionId, callback) => metadataGetObject(bucketName, entry.key,
-                versionId, log, (err, objMD) => {
-                    // if general error from metadata return error
-                    if (err) {
-                        monitoring.promMetrics('DELETE', bucketName, err.code,
-                            'multiObjectDelete');
-                        return callback(err);
-                    }
-                    if (!objMD) {
-                        const verCfg = bucket.getVersioningConfiguration();
-                        // To adhere to AWS behavior, create a delete marker
-                        // if trying to delete an object that does not exist
-                        // when versioning has been configured
-                        if (verCfg && !entry.versionId) {
-                            log.debug('trying to delete specific version ' +
-                            ' that does not exist');
-                            return callback(null, objMD, versionId);
+    return async.waterfall([
+        callback => initialStep((err, cache) => callback(err, cache)),
+        (cache, callback) => async.forEachLimit(inPlay, 50, (entry, moveOn) => {
+            async.waterfall([
+                callback => callback(...processObjectVersion(entry, bucketName)),
+                // for obj deletes, no need to check acl's at object level
+                // (authority is at the bucket level for obj deletes)
+                (versionId, callback) => metadataGetObject(bucketName, entry.key,
+                    versionId, log, (err, objMD) => {
```

Comment:
should just pass 'callback' and move this anonymous function as the next stage in the `async.waterfall`

Reply:
What is the rationale? That is not what was done before, and I do not want to change the structure...

Reply:
Done: https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-e8319ec288d282012f389d259d6468a0922b91ec7cc52f2b8c15b01ff3ba9c2aR260

Reply:
rationale is really just that this is a waterfall call : whose call is to replace the "final" callback by the next stage in waterfall...

ie instead of:
```
metadataGetObject(bucketName, entry.key, versionId, log, (err, objMD) => {
    [...]
```
we should have:
```
metadataGetObject(bucketName, entry.key, versionId, log, callback),
(err, objMD) => {
    [...]
```

(unless there are other subtleties, there may be reasons why we cannot do this...)